File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) 
    {
        int l = 0;
        int r = height.size()-1;
        int water = 0;
        int Lmax = height[l];
        int Rmax = height[r];

        while ( l <= r)
        {
            if ( Lmax < Rmax)
            {
                if ( Lmax < height[l]) Lmax = height[l];
                water += Lmax - height[l];
                l++;
            }
            else if ( Rmax <= Lmax)
            {
                if ( Rmax < height[r]) Rmax = height[r];
                water += Rmax - height[r];
                r--;
            }
        }
        return water;
        
    }
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        Lmax = [0] * n
        Lmax[0] = height[0]
        lmx = Lmax[0]

        for i in range(1,n):
            if height[i] > lmx:
                lmx = height[i]
            Lmax[i] = lmx

        Rmax = [0] * n
        Rmax[-1] = height[-1] 
        rmx = Rmax[-1]

        for i in range(n-2,-1,-1):
            if height[i] > rmx:
                rmx = height[i]
            Rmax[i] = rmx

        ans = 0

        for i in range(0,n):
            water  = min(Lmax[i],Rmax[i]) - height[i]
            ans += water

        return ans

        